components/kserve/kserve.go

```diff
@@ -144,6 +152,13 @@ func (k *Kserve) ReconcileComponent(ctx context.Context, cli client.Client, resC
 			return err
 		}
 	}
+
+	if enabled {
+		if err := k.setupKserveConfigAndDependencies(ctx, cli, dscispec); err != nil {
```

Comment:
This effectively moves Serverless setup to the end of the KServe reconcile cycle.
I'm not sure that's 100% correct because KServe pods may the have a chance to start before KNative Serving is fully installed, which may lead to KServe to boot-up in non-Serverless mode.

Reply:
Additionally, taking into account your changes in opendatahub-io/odh-model-controller#164, I don't think this should be that late in the reconcile cycle.

An upgrade may lead to `odh-model-controller` seeing the wrong values of `defaultDeploymentMode` and lead to reconcile issues. So, right ordering seems to be important:
1. Make sure Serverless is installed, before trying to install KServe.
2. Deploy KServe
3. Make sure the ConfigMap is patched with the right deployment mode before deploying odh-model-controller.
4. Deploy odh-model-controller.

Bullet 4 is not trivial, because ModelMesh may also try to deploy it out-of-order.

Reply:
Also, thinking this together with kserve/kserve#3472:

by updating the ConfigMap after KServe is deployed, there is the chance that KServe boots before the ConfigMap is patched. Because of this, as part of bullet 3, I think you will need to restart the KServe pod (if it already exists) to ensure it will read the updated deployment mode.

Reply:
Changed the logic in the latest commit. 
- Moved Serverless setup to earlier in the reconcile loop before Kserve is deployed. 
- Moved Kserve Config setup before odh model controller deployment 
- If the configmap is updated during the reconcile loop, the kserve pod is restarted to reboot with values from the updated configmap
  - If the reconcile loop is triggered but the configmap is not updated, the kserve pod is not restarted

Reply:
i agree the first 2 points except the last one to have a "kill pod => restart it" inside of this function. 

in this case, the feature is tightly coupled with odh operator.
e.g user can only switch deployment by using this flag via DSC.kserve, (if manually update configmap data, it wont take effects, unless user restart pod later by themselves).
i can accept this as a workaround done by odh operator for now, but in long term, i would think this "kill pod => restart it" logic should be something done by the kserve controller.

Reply:
> but in long term, i would think this "kill pod => restart it" logic should be something done by the kserve controller.

I'm not sure if this would be doable, because KServe controller does not have awareness of odh-model-controller. The odh-model-controller is becoming like a glue component for ODH platform.

IMHO, even moving the restart to another controller would be a workaround. What I see more ideal is to re-work the patch as a transformer (which could be a generalized feature in the operator), similar to the Namespace and Labels transformers you already have. Manifests would be modified accordingly and the `manageResource` func would use the final resources, rather than patching after applying. This would fully remove the need of the pod restart.

Reply:
> 
>  What I see more ideal is to re-work the patch as a transformer (which could be a generalized feature in the operator), similar to the Namespace and Labels transformers you already have. Manifests would be modified accordingly and the `manageResource` func would use the final resources, rather than patching after applying. This would fully remove the need of the pod restart... but, well, I think this would be a enhancement for another PR - we could coordinate to make this near term.

And that was originally proposed :)

Reply:
@ykaliuta Ah, interesting! I guess it is related to [this comment](https://github.com/opendatahub-io/opendatahub-operator/pull/864#issuecomment-1943582932), right?

I didn't find your original comment, so I didn't have the full context and never followed up. Well... you can blame me about that :sweat_smile:

Reply:
> @ykaliuta Ah, interesting! I guess it is related to [this comment](https://github.com/opendatahub-io/opendatahub-operator/pull/864#issuecomment-1943582932), right?
> 
> I didn't find your original comment, so I didn't have the full context and never followed up. Well... you can blame me about that 😅

Right, but one comment above, this https://github.com/opendatahub-io/opendatahub-operator/pull/864#issuecomment-1943582932 and originally it was a conversation off-thread. No blaming at all. The conclusion came there.  So there are pros and cons https://github.com/opendatahub-io/opendatahub-operator/pull/864#issuecomment-1944124092